proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1144 @@
+  to have.)
+
+* ``Experimental2024``: Switching to the ``Experimental`` series of
+  editions gives you access to experimental features of GHC, which might
+  reasonably evolve and break existing programs.
+
+* ``Latest``: This language edition is the latest and greatest that GHC
+  has to offer. Compiling with ``Latest`` might break between releases.
+
+* ``Student2024``: The student edition of the language has extra guardrails.
+  Furthermore, any code written in the ``Student`` edition is expected
+  to compile with all future versions of GHC for 10 years, meant to
+  echo the expected lifetime of a textbook.
+
+With that out of the way, here are some scenarios that the language editions
+model helps us to address:
+
+* **Stability**. Once we include a feature in a ``Stable`` edition, it
+  will not change. However, that does not mean GHC cannot evolve. If we
+  identify a ``Stable`` feature that nevertheless deserves an upgrade, we
+  can do so, provided we continue to support the old behavior. So
+  type inference around ``FlexibleInstances``, say, might work one way in
+  ``Stable2024`` and differently in ``Stable2027``. If we release a new
+  ``Stable`` edition every 3 years, users have three years to upgrade before
+  we no longer guarantee support. A three-year timeline is long enough that
+  we might imagine contributors writing upgrade tools in that timeframe.
+
+* **Error messages**. Because the language edition describes the user (very
+  coarsely!) we can tailor error messages to be appropriate for them. The
+  ``Student`` editions will not advise ``DataKinds``. The ``Stable`` editions
+  will not advise ``LinearTypes``.
+
+* **Warning evolution**. One challenge in adding features to GHC is to
+  know how to evolve the warning system. Does a new warning get added to
+  ``-Wall``? And what's the deal with ``-Wcompat``?
+
+  Language editions make
+  these questions easier to answer: If we think that a warning should be
+  enabled for users going forward, we can turn it on by default, but only
+  in appropriate language editions, such as ``Experimental`` or ``Latest``.
+
+  For ``-Wcompat``, we add warnings that describe features that will change
+  in the next edition in a given series. So the ``Stable2024`` edition of
+  ``-Wcompat`` will warn about features changing in ``Stable2027``. This
+  set of warnings will likely grow between 2024 and 2027; that's fine.
+  (We do *not* guarantee that all ``Stable`` code remain warning-free, especially
+  with ``-Wcompat`` enabled.) Now if a user is running ``Stable2024`` in
+  2028 with ``-Wcompat``, they won't see warnings about changes due in
+  2030; they'll see just the warnings they need to upgrade to ``Stable2027``.
+
+* **Simplicity**. The goal is that a vast majority of our users will be
+  able to specify a language edition, and that's it. No extensions. No
+  warning flags. This simplifies what a user needs to think about when
```

Comment:
Optimistic: yes. But I actually think it would be interesting and worthwhile to understand what warning flags are needed. That is, enabling a warning moves responsibility from the programmer to the compiler (e.g. the programmer no longer needs to worry about the completeness of their patterns), or establishes some property of a program (like the ability to find a binding site of a variable using a simple analysis). So if users are enabling more warnings, I'd want to understand what property they're worried about, so we can perhaps design our system better. That said, we probably won't do this (it's hard). I'm fine with warning flags. I just think that each warning flag is a small message that we've slightly mis-designed something.

Reply:
>  So if users are enabling more warnings, I'd want to understand what property they're worried about, so we can perhaps design our system better.

This sounds good in the abstract but I'm not sure I buy it in the concrete. Consider unused variable warnings. Users want them on when they want to spot any little thing that might be wrong with their code, or tidy it up for others to read. Or, at other times they may not care about such things and want to turn them off. I don't see a way to design the system better here: users just _care about different things at different times_.

Reply:
I agree with @michaelpj here. Warning flags are useful because they allow the compiler to offer a menu of different analyses with different trade-offs, so the user can choose which they need at a particular point in time (which may well change depending on where they are in the development cycle of a particular program, or just depending on what they are currently interested in doing). This includes warnings about essentially stylistic or otherwise controversial choices (e.g. `Wx-partial`), where the alternative is for the compiler authors to be forced to pick a side in the controversy (either by making a warning mandatory or by not supporting it).